inference-engine/thirdparty/clDNN/src/program.cpp

```diff
@@ -1309,4 +1312,37 @@ void program_impl::set_layout_optimizer_attributes(layout_optimizer& lo) {
 
     if (should_use_bs_fs_yx_bsv16_fsv16)
         lo.set_optimization_attribute(layout_optimizer::optimization_attributes_type::bs_fs_yx_bsv16_fsv16_network, 1);
+
+
+    // This is to avoid using fsv16 for shallow-feature networks.
+    // This may not be exactly same as real execution graph as layer fusing is not done yet,
+    // but it is a reasonable approximation.
+    // Check the expected network efficiency after setting layer optimization attributes.
+    // If network depth is shallow, it is faster with fsv4.
+    for (auto& node : get_processing_order()) {
+        auto &prim = *node;
+
+        if (prim.is_in_data_flow() && prim.type() == cldnn::convolution::type_id()) {
+            size_t num_feature = prim.get_output_layout().size.feature.vector()[0];
+            size_t num_spatial = 1;
+            for (auto s : prim.get_output_layout().size.spatial.vector())
+                num_spatial *= s;
+
+            if (lo.get_preferred_format(prim) != format::b_fs_yx_fsv4) {
+                weight_sum += num_spatial;
```

Comment:
I wonder what will happen if a model has 50% of fp16 convolution and 50% of int8 convolutions with shallow feature depth? I assume that fsv16 kernels will be disabled and fsv4 will be used for int8 kernels, and bfyx format will be used for fp16 ones which seems to be inefficient. So don't we need an extra condition that will ensure that most of the convolutions in a model are quantized?

Reply:
I did not take such scenario into consideration. In such cases, the efficient layout would be `fsv4 for int8` and `fsv16 for fp16`? Could you share such network if you have so that I can test?

It was not trivial to set the criteria. I placed this code with the `fsv4` check condition because of the scenario like below:
```
conv1 : fsv16, depth 16
conv2 : fsv16, depth 16
conv3 : fsv4, depth 1
conv4 : fsv4, depth 3
conv5 : fsv4, depth 3
```
For such network, we need to turn `fsv16` on, as some layers run in `fsv4` though `should_fsv16=true`. However, the average becomes quite low because of conv3,4,5.
Because of such scenario, 1) I moved this code after initial layout selection 2) in order to exclude the layers with fsv4.

I think this is a structural issue with current network optimization. With hand-written layout optimization, it was not trivial to find a strong metric :(